input/amqp.go

```diff
@@ -144,12 +144,62 @@ type amqpConnector func(a *Amqp) error
 // AMQPConnector connects using the given configuration
 func AMQPConnector(a *Amqp) error {
 	log.Infof("dialing AMQP: %v", a.uri)
-	conn, err := amqp.Dial(a.uri.String())
-	if err != nil {
-		return err
+
+	var conn *amqp.Connection
+	var err error
+
+	for {
+		config := amqp.Config{
+			Heartbeat: time.Duration(a.config.Amqp.Amqp_heartbeat) * time.Second,
+		}
+
+		conn, err = amqp.DialConfig(a.uri.String(), config)
+		if err == nil {
+			break
+		}
+
+		if !a.config.Amqp.Amqp_retry {
+			return err
+		}
+
+		log.Errorf("Failed to connect to AMQP server: %v. Retrying in %d seconds...", err, a.config.Amqp.Amqp_retrydelay)
+		time.Sleep(time.Duration(a.config.Amqp.Amqp_retrydelay) * time.Second)
 	}
+
 	a.conn = conn
 
+	// Create a channel to receive close notifications from the connection
+	closeCh := make(chan *amqp.Error)
+	conn.NotifyClose(closeCh)
+
+	// Start a goroutine to monitor the connection state
+	go func() {
+		for {
+			select {
+			case <-closeCh:
+				log.Println("AMQP connection closed.")
+
+				if !a.config.Amqp.Amqp_retry {
+					log.Println("Retry is disabled. Exiting reconnection attempt.")
+					return
+				}
+
+				log.Println("Attempting to reconnect...")
+
+				for {
+					err := AMQPConnector(a)
+					if err == nil {
+						log.Println("Successfully reconnected to AMQP server.")
+						return
```

Comment:
Shouldn't this be a `break`?

Reply:
Never mind, I didn't notice the recursion.

Reply:
Thanks for checking @npazosmendez! In this context, return is used to exit the goroutine once the reconnection is successful. If we use break instead of return, it would only exit the inner for loop, but the outer for loop (which is monitoring the connection state) would continue running. This is not the desired behavior because we want to stop the reconnection attempts and exit the goroutine once the reconnection is successful.

Reply:
Makes sense. I did notice something else, starting a new thread